File: frontend/src/features/crm/types/crm.ts
export interface PipelineStage {
  id: string;
  tenant_id: string;
  name: string;
  order: number;
  probability: number;
  is_won: boolean;
  is_lost: boolean;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface Lead {
  id: string;
  tenant_id: string;
  name: string;
  company?: string | null;
  email?: string | null;
  phone?: string | null;
  whatsapp?: string | null;
  job_title?: string | null;
  website?: string | null;
  source: string;
  status: string;
  priority: string;
  industry?: string | null;
  estimated_value: number;
  assigned_user_id?: string | null;
  follow_up_date?: string | null;
  notes?: string | null;
  lost_reason?: string | null;
  address?: string | null;
  city?: string | null;
  state?: string | null;
  country?: string | null;
  created_at: string;
  updated_at: string;
  deleted_at?: string | null;
}

export interface LeadCreate {
  name: string;
  company?: string;
  email?: string;
  phone?: string;
  whatsapp?: string;
  job_title?: string;
  website?: string;
  source?: string;
  status?: string;
  priority?: string;
  industry?: string;
  estimated_value?: number;
  assigned_user_id?: string;
  follow_up_date?: string;
  notes?: string;
  lost_reason?: string;
  address?: string;
  city?: string;
  state?: string;
  country?: string;
}

export type LeadUpdate = Partial<LeadCreate>;

export interface PaginatedLeads {
  items: Lead[];
  total: number;
  page: number;
  limit: number;
}

export interface Deal {
  id: string;
  tenant_id: string;
  title: string;
  value: number;
  currency: string;
  stage_id: string;
  stage?: PipelineStage | null;
  lead_id?: string | null;
  customer_id?: string | null;
  probability: number;
  expected_closing_date?: string | null;
  actual_closing_date?: string | null;
  owner_id?: string | null;
  won_reason?: string | null;
  lost_reason?: string | null;
  notes?: string | null;
  created_at: string;
  updated_at: string;
  deleted_at?: string | null;
}

export interface DealCreate {
  title: string;
  value?: number;
  currency?: string;
  stage_id: string;
  lead_id?: string;
  customer_id?: string;
  probability?: number;
  expected_closing_date?: string;
  owner_id?: string;
  notes?: string;
}

export interface DealUpdate extends Partial<DealCreate> {
  won_reason?: string;
  lost_reason?: string;
  actual_closing_date?: string;
}

export interface PaginatedDeals {
  items: Deal[];
  total: number;
  page: number;
  limit: number;
}

export interface Activity {
  id: string;
  tenant_id: string;
  type: string; // Call, Meeting, Email, WhatsApp, Note, Task
  subject: string;
  description?: string | null;
  due_date?: string | null;
  completed_at?: string | null;
  status: string; // pending, completed
  priority: string;
  assigned_user_id?: string | null;
  created_by_id?: string | null;
  lead_id?: string | null;
  deal_id?: string | null;
  customer_id?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ActivityCreate {
  type?: string;
  subject: string;
  description?: string;
  due_date?: string;
  status?: string;
  priority?: string;
  assigned_user_id?: string;
  lead_id?: string;
  deal_id?: string;
  customer_id?: string;
}

export interface ActivityUpdate extends Partial<ActivityCreate> {
  completed_at?: string;
}

export interface Customer {
  id: string;
  tenant_id: string;
  name: string;
  company?: string | null;
  email?: string | null;
  phone?: string | null;
  whatsapp?: string | null;
  converted_from_lead_id?: string | null;
  assigned_user_id?: string | null;
  created_at: string;
  updated_at: string;
}

File: backend/app/schemas/crm.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, Field


class PipelineStageResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    name: str
    order: int
    probability: int
    is_won: bool
    is_lost: bool
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class LeadBase(BaseModel):
    name: str = Field(..., min_length=1)
    company: str | None = None
    email: EmailStr | None = None
    phone: str | None = None
    whatsapp: str | None = None
    job_title: str | None = None
    website: str | None = None

    source: str = "Website"
    status: str = "New"  # New, Contacted, Qualified, Converted, Lost
    priority: str = "Medium"  # Low, Medium, High
    industry: str | None = None
    estimated_value: int = 0

    assigned_user_id: UUID | None = None
    follow_up_date: datetime | None = None
    notes: str | None = None
    lost_reason: str | None = None

    address: str | None = None
    city: str | None = None
    state: str | None = None
    country: str | None = None


class LeadCreate(LeadBase):
    pass


class LeadUpdate(BaseModel):
    name: str | None = None
    company: str | None = None
    email: EmailStr | None = None
    phone: str | None = None
    whatsapp: str | None = None
    job_title: str | None = None
    website: str | None = None

    source: str | None = None
    status: str | None = None
    priority: str | None = None
    industry: str | None = None
    estimated_value: int | None = None

    assigned_user_id: UUID | None = None
    follow_up_date: datetime | None = None
    notes: str | None = None
    lost_reason: str | None = None

    address: str | None = None
    city: str | None = None
    state: str | None = None
    country: str | None = None


class LeadResponse(LeadBase):
    id: UUID
    tenant_id: UUID
    created_at: datetime
    updated_at: datetime
    converted_at: datetime | None = None
    converted_customer_id: UUID | None = None
    deleted_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class CustomerCreate(BaseModel):
    name: str = Field(..., min_length=1)
    phone: str = Field(..., min_length=1)
    email: str | None = None
    customer_type: str = "Individual"
    company: str | None = None
    billing_address: str | None = None
    city: str | None = None
    state: str | None = None
    pincode: str | None = None
    gstin: str | None = None
    pan: str | None = None
    opening_balance: float = 0.0
    opening_balance_type: str = "Debit"
    credit_limit: float | None = None
    notes: str | None = None
    whatsapp: str | None = None
    assigned_user_id: UUID | None = None


class CustomerUpdate(BaseModel):
    name: str | None = None
    phone: str | None = None
    email: str | None = None
    customer_type: str | None = None
    company: str | None = None
    billing_address: str | None = None
    city: str | None = None
    state: str | None = None
    pincode: str | None = None
    gstin: str | None = None
    pan: str | None = None
    opening_balance: float | None = None
    opening_balance_type: str | None = None
    credit_limit: float | None = None
    notes: str | None = None
    whatsapp: str | None = None
    assigned_user_id: UUID | None = None


class CustomerResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    name: str
    company: str | None = None
    email: str | None = None
    phone: str | None = None
    whatsapp: str | None = None
    customer_type: str = "Individual"
    city: str | None = None
    state: str | None = None
    pincode: str | None = None
    gstin: str | None = None
    pan: str | None = None
    billing_address: str | None = None
    opening_balance: float = 0.0
    opening_balance_type: str = "Debit"
    credit_limit: float | None = None
    notes: str | None = None
    converted_from_lead_id: UUID | None = None
    assigned_user_id: UUID | None = None
    outstanding_balance: float = 0.0
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class PaginatedCustomersResponse(BaseModel):
    items: list[CustomerResponse]
    total: int
    page: int
    limit: int


class CustomerImportRowError(BaseModel):
    row: int
    reason: str


class CustomerImportSummary(BaseModel):
    imported_count: int
    skipped_count: int
    failed_count: int
    errors: list[CustomerImportRowError] = []


class DealBase(BaseModel):
    title: str = Field(..., min_length=1)
    value: int = 0
    currency: str = "USD"
    stage_id: UUID
    lead_id: UUID | None = None
    customer_id: UUID | None = None
    probability: int = 0
    expected_closing_date: datetime | None = None
    actual_closing_date: datetime | None = None
    owner_id: UUID | None = None
    won_reason: str | None = None
    lost_reason: str | None = None
    notes: str | None = None


class DealCreate(DealBase):
    pass


class DealUpdate(BaseModel):
    title: str | None = None
    value: int | None = None
    currency: str | None = None
    stage_id: UUID | None = None
    lead_id: UUID | None = None
    customer_id: UUID | None = None
    probability: int | None = None
    expected_closing_date: datetime | None = None
    actual_closing_date: datetime | None = None
    owner_id: UUID | None = None
    won_reason: str | None = None
    lost_reason: str | None = None
    notes: str | None = None


class DealResponse(DealBase):
    id: UUID
    tenant_id: UUID
    stage: PipelineStageResponse | None = None
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class ActivityBase(BaseModel):
    type: str = "Note"  # Call, Meeting, Email, WhatsApp, Note, Task
    subject: str = Field(..., min_length=1)
    description: str | None = None
    due_date: datetime | None = None
    completed_at: datetime | None = None
    status: str = "pending"  # pending, completed
    priority: str = "Medium"
    assigned_user_id: UUID | None = None

    lead_id: UUID | None = None
    deal_id: UUID | None = None
    customer_id: UUID | None = None


class ActivityCreate(ActivityBase):
    pass


class ActivityUpdate(BaseModel):
    type: str | None = None
    subject: str | None = None
    description: str | None = None
    due_date: datetime | None = None
    completed_at: datetime | None = None
    status: str | None = None
    priority: str | None = None
    assigned_user_id: UUID | None = None


class ActivityResponse(ActivityBase):
    id: UUID
    tenant_id: UUID
    created_by_id: UUID | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PaginatedLeadsResponse(BaseModel):
    items: list[LeadResponse]
    total: int
    page: int
    limit: int


class PaginatedDealsResponse(BaseModel):
    items: list[DealResponse]
    total: int
    page: int
    limit: int
